packages/next/src/utils.ts

```diff
@@ -2677,3 +2679,49 @@ function transformSourceMap(
 
   return { ...sourcemap, sources };
 }
+
+interface LambdaGroupTypeInterface {
+  isApiLambda: boolean;
+  isPrerenders?: boolean;
+}
+
+export function getOperationType({
+  group,
+  prerenderManifest,
+  pageFileName,
+}: {
+  group?: LambdaGroupTypeInterface;
+  prerenderManifest?: NextPrerenderedRoutes;
+  pageFileName?: string;
+}) {
+  if (group?.isApiLambda || isApiPage(pageFileName)) {
+    return 'API';
+  }
+
+  if (group?.isPrerenders) {
+    return 'ISR';
+  }
+
+  if (pageFileName) {
+    const routePaths = Object.keys({
+      ...prerenderManifest?.blockingFallbackRoutes,
+      ...prerenderManifest?.fallbackRoutes,
+    });
```

Comment:
How many `routePaths` do we expect here? Is it bounded or can it grow to N?

It could be a lot faster to do an early return like this:

```ts
for (const key of Object.keys(blockingFallbackRoutes)) {
  if (key === pageFileName) return 'ISR'
}

for (const key of Object.keys(fallbackRoutes)) {
  if (key === pageFileName) return 'ISR'
}
```

Reply:
If the same `prerenderManifest` is passed to this function more than once, its probably best to hoist it up and precompute a Set that is then passed into this function so you can do `set.has(pageFileName)`

Reply:
Yeah, I thought about commenting on this too, but didn't think of breaking it into two different tests. Maybe skip the `Object.keys()` call and  allocation and use `Object.prototype.hasOwnProperty.call(blockingFallbackRoutes, pageFilename)`?

Reply:
Done! [`ecf4712` (#9196)](https://github.com/vercel/vercel/pull/9196/commits/ecf4712d8ea8ed352620e625caa55a00f070ec3d)

Reply:
I think precomputing the Set is actually best. See existing implementation here:

https://github.com/vercel/vercel/blob/ecf4712d8ea8ed352620e625caa55a00f070ec3d/packages/next/src/index.ts#L501-L503

Reply:
For performance, `Object.prototype.hasOwnProperty.call()` is going to be faster and more efficient since there's no memory being allocated, but it does look ugly.

```
if (pageFileName) {
  const {
    blockingFallbackRoutes = {},
    fallbackRoutes = {},
  } = prerenderManifest || {};
  const hasOwn = Object.prototype.hasOwnProperty;
  if (hasOwn.call(blockingFallbackRoutes, pageFileName) || hasOwn.call(fallbackRoutes, pageFileName)) {
    return 'ISR';
  }
}
```

How about using `in`?

```
if (pageFileName) {
  const {
    blockingFallbackRoutes = {},
    fallbackRoutes = {},
  } = prerenderManifest || {};
  if (pageFileName in blockingFallbackRoutes || pageFileName in fallbackRoutes) {
    return 'ISR';
  }
}
```

There used to be performance issues with `in`, but it seems to be ok these days.

If sticking with `Object.keys()`, then perhaps `if (Object.keys(<obj>).find(k => k === pageFileName)) { }`?

Reply:
Yeah “in” could work too 👍

Reply:
I did some [quick perf testing](https://gist.github.com/EndangeredMassa/89f4ea35d5365537c7cb1d1f22a42595). Let me know if any of that looks wrong. If correct:

- looping with short circuiting: slow
- in operator: very fast
- building up a set to use: very slow
- checking that set: very fast

Yeah, `in` looks like a win here. [`184ef3d` (#9196)](https://github.com/vercel/vercel/pull/9196/commits/184ef3d6b5185b3068a74d22d7cf57d668058e46)